src/ir/ir_traversing_visitor.h

```diff
@@ -24,47 +26,89 @@ namespace raksha::ir {
 // A visitor that also traverses the children of a node and allows performing
 // different actions before (PreVisit) and after (PostVisit) the children are
 // visited. Override any of the `PreVisit` and `PostVisit` methods as needed.
-template <typename Derived>
-class IRTraversingVisitor : public IRVisitor<Derived> {
+template <typename Derived, typename Result = Unit>
+class IRTraversingVisitor : public IRVisitor<Derived, Result> {
  public:
   virtual ~IRTraversingVisitor() {}
 
+  // Gives a default value for all 'PreVisit's to start with.
+  // Should be over-ridden if the Result is not default constructable.
+  virtual Result GetDefaultValue() { return Result(); }
```

Comment:
This looks good, but you may run into compilation errors if this is instantiated with a `Result` type without a default constructor.

It looks like overriding `GetDefaultValue` is currently untested. I think you should make a (very simple) visitor to try out returning a type with the no-arg constructor explicitly deleted. So something like:

```C++
class WrappedInt {
  public:
  WrappedInt() = delete;
  WrappedInt(uint64_t val) : val_(val) {}
  
  int64_t operator*() const {
    return val;
  }
  
  private:
  uint64_t val_;
};
```

And just make a visitor using that to count the nodes in the tree.

Reply:
If you do run into compilation errors with this default `GetDefaultValue`, you can still have a nice default like you have here, but unfortunately, I think getting it may rely on dark template magic:

```C++
template<class Result, class Enable>
struct DefaultValueGetter {
  Result Get() { static_assert("Default value override is required!"); }
}

template<class Result>
struct DefaultValueGetter<Result, std::enable_if_t<std::is_default_constructible_v<Result>>> {
  Result Get() { return Result(); }
};

virtual Result GetDefaultValue() { return DefaultValueGetter<Result>::Get(); }
```

Reply:
Ah, I see now, because the template generates the functions for the base class (even though they are shadowed by the `override` we have to provide a 'working' default constructor in this case. That is a pretty bad place for C++ to be in./
I'm sorry that I hadn't foreseen it.

I'll use this template approach, thanks!